Implementation of product tests

MariuszSolvd opened the pull request.

MariuszSolvd committed:
Implementation of LoginService

MariuszSolvd committed:
Naming adjustment

MariuszSolvd committed:
Refactoring header name

MariuszSolvd committed:
Added product list

MariuszSolvd committed:
Added Product

MariuszSolvd committed:
Added clickOnProduct

MariuszSolvd committed:
Added ProductDetailPageBase and ProductDetailPage

MariuszSolvd committed:
Added Product entity

MariuszSolvd committed:
Implemented 1st test for product page

MariuszSolvd committed:
Added isOpened methods, and assertion to test

MariuszSolvd committed:
Added SortBase

MariuszSolvd committed:
Added SortButton

MariuszSolvd committed:
Added Sort

MariuszSolvd committed:
Added isOpened()

MariuszSolvd committed:
Implemented 2nd test for product page

MariuszSolvd committed:
Implemented 3rd test for product page

MariuszSolvd committed:
Change price type in product related objects from String to double, to avoid sort problems

MariuszSolvd committed:
Implemented 4th test

MariuszSolvd committed:
Implemented 5th test

MariuszSolvd committed:
Added Remove button to Product Objects

MariuszSolvd committed:
Added getters for remove&add to cart button

MariuszSolvd committed:
Implemented 6th test

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductDetailPage.java:
```diff
@@ -0,0 +1,88 @@
+    private ExtendedWebElement image;
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeOther[`name == \"test-Description\"`]/*[1]")
+    private ExtendedWebElement title;
```
```suggestion
    private ExtendedWebElement titleLabel;
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductDetailPage.java:
```diff
@@ -0,0 +1,88 @@
+    private ExtendedWebElement title;
+
+    @ExtendedFindBy(iosPredicate = "name == \"test-Price\"")
+    private ExtendedWebElement price;
```
```suggestion
    private ExtendedWebElement priceLabel;
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/base/elements/ProductBase.java:
```diff
@@ -0,0 +1,32 @@
+
+    public abstract ProductDetailPageBase clickOnProduct();
+
+    public abstract String getImage();
```
could you please explain what do you get here? if you get image URL then I suggest to rename method to getImageUrl(). please specify what exactly you get, it makes support of the code easier

MariuszSolvd replied:
Yes, it's image url. I will rename it.

ankochergin commented on src/main/java/com/solvd/pages/base/elements/ProductBase.java:
```diff
@@ -0,0 +1,32 @@
+
+    public abstract double getPrice();
+
+    public abstract ExtendedWebElement getAddToCartButton();
```
lease never return elements from the page object, it's a bad practice. the idea of page object pattern is to incapsulate all elements and interactions with then inside of single class

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductPage.java:
```diff
@@ -1,22 +1,33 @@
+    private ExtendedWebElement sortButton;
+
+    @ExtendedFindBy(iosPredicate = "name == \"test-Item\"")
+    private List<Product> products;
```
I don't recommend to use this approach for mobile tests because you cannot use Base class here, please see comments in Product class

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductPage.java:
```diff
@@ -1,22 +1,33 @@
+
     public ProductPage(WebDriver driver) {
         super(driver);
     }
 
+
     @Override
     public boolean isOpened() {
```
```suggestion
    }

    @Override
    public boolean isOpened() {
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/ios/elements/Product.java:
```diff
@@ -0,0 +1,74 @@
+    public Product(WebDriver driver, SearchContext searchContext) {
+        super(driver, searchContext);
+    }
+
```
I recommend to avoid AbstractUIObject for mobile tests.instead of it extend it from regular AbstractPage and create methods like `clickProductTitle(String productName)` (preferable) or `clickProductTitle(int index)`
and format locator using this argument in methods by product title or index

MariuszSolvd committed:
Fixing naming

MariuszSolvd committed:
Removed field List<Product>

MariuszSolvd committed:
Change Product from AbstractUIObject, to AbstractPage

MariuszSolvd committed:
Added countProduct method

MariuszSolvd committed:
Adjusted 1st test to changes

MariuszSolvd committed:
Adjusted 2nd test to changes

MariuszSolvd committed:
Small fix in countProduct()

MariuszSolvd committed:
Adjusted 3rd test to changes

MariuszSolvd committed:
Adjusted 4th test to changes

MariuszSolvd committed:
Adjusted 5th test to changes

MariuszSolvd committed:
Change name of int var

MariuszSolvd committed:
Remove of isProductPresent() method

MariuszSolvd committed:
Add check if button is present

MariuszSolvd committed:
Change from ExtendedWebElement to boolean return type

MariuszSolvd committed:
Adjust 6th test to changes
Small fixes in 1st test

ankochergin commented on src/main/java/com/solvd/pages/base/ProductPageBase.java:
```diff
@@ -1,10 +1,15 @@
+
+    public abstract SortBase clickOnSortButton();
+
+    public abstract int countProducts();
```
```suggestion
    public abstract int getProductsCount();
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/base/elements/SortBase.java:
```diff
@@ -0,0 +1,21 @@
+import com.solvd.pages.base.ProductPageBase;
+import org.openqa.selenium.WebDriver;
+
+public abstract class SortBase extends CommonPageBase {
```
```suggestion
public abstract class SortBoxBase extends CommonPageBase {
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductPage.java:
```diff
@@ -1,18 +1,26 @@
+
     public ProductPage(WebDriver driver) {
         super(driver);
     }
```
```suggestion
    @ExtendedFindBy(iosPredicate = "name == \"test-Modal Selector Button\"")
    private ExtendedWebElement sortButton;
    
    @ExtendedFindBy(iosPredicate = "name == \"test-Item\"")
    private List<ExtendedWebElement> productContainers;

    public ProductPage(WebDriver driver) {
        super(driver);
    }
```

MariuszSolvd replied:
Done

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductPage.java:
```diff
@@ -26,4 +34,16 @@ public boolean isOpened() {
+
+    @Override
+    public int countProducts() {
+        return findExtendedWebElements(AppiumBy.iOSNsPredicateString("name == \"test-Item\"")).size() + 1;
```
```suggestion
        return productContainers.size() + 1;
```

MariuszSolvd replied:
Done

MariuszSolvd committed:
Rename to SortBox

MariuszSolvd committed:
Add productContainer

MariuszSolvd committed:
Adjusted changes in tests

MariuszSolvd committed:
Added Sorter
SortBox has now one method(Need to pass Sort object as an argument)
Test Adjustments

ankochergin commented on src/main/java/com/solvd/utilis/enums/Sorter.java:
```diff
@@ -0,0 +1,9 @@
+
+public enum Sorter {
+
+    A_TO_Z,
```
you can create a String parameter for each sort option and in this case you will be able to use it as a parameter for your locator
```suggestion
    A_TO_Z("Name (A to Z)"),
```

ankochergin commented on src/main/java/com/solvd/pages/ios/elements/SortBox.java:
```diff
@@ -0,0 +1,49 @@
+
+    public SortBox(WebDriver driver) {
+        super(driver);
+    }
+
+
+    @Override
+    public boolean isOpened() {
+        return isOpened(IS_OPENED_DEFAULT_TIMEOUT);
```
```suggestion
    }

    @Override
    public boolean isOpened() {
        return isOpened(IS_OPENED_DEFAULT_TIMEOUT);
```

ankochergin commented on src/main/java/com/solvd/pages/ios/ProductDetailPage.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+
+}
```
```suggestion
    }

}
```

MariuszSolvd committed:
Now using one field for sort, and passing classChain as an argument by Sort

MariuszSolvd committed:
Add Products Test to testng.xml

MariuszSolvd committed:
Deleted one empty line
